cpp/memilio/compartments/flow_model.h

```diff
@@ -0,0 +1,280 @@
+ *
+ * Inherits from  @see CompartmentalModel, and defines the derivatives depending on the flows. Hence, a model implementing
+ * FlowModel has to define the function get_flows, instead of get_derivatives.
+ *
+ * Flows is expected to be a TypeList containing types Flow<A,B>, where A and B are compartments from the enum Comp.
+ * Some examples can be found in the cpp/models/ directory, within the model.h files.
+ */
+template <class Comp, class Pop, class Params, class Flows>
+struct FlowModel : public CompartmentalModel<Comp, Pop, Params> {
+    using PopIndex = typename Pop::Index;
+    // FlowIndex is the same as PopIndex without the category Comp. It is used as argument type for
+    // get_flat_flow_index, since a flow is used to determine only the compartment (i.e. Index<Comp>) for the
+    // population of the model.
+    // The remaining indices (those contained in FlowIndex) must still be provided to get an entry of population.
+    // This approach only works with exactly one category of type Comp in PopIndex (hence the assertion below).
+    using FlowIndex = details::filtered_index_t<Comp, Index, PopIndex>;
+    // Enforce that Comp is a unique Category of PopIndex, since we use Flows (via their source/target) to provide
+    // Comp indices for the population.
+    static_assert(FlowIndex::size == PopIndex::size - 1, "Compartments must be used exactly once as population index.");
+
+public:
+    using Base = CompartmentalModel<Comp, Pop, Params>;
+    /**
+     * @brief Default constructor, forwarding args to Base constructor.
+     */
+    template <class... Args>
+    FlowModel(Args... args)
+        : CompartmentalModel<Comp, Pop, Params>(args...)
+        , m_flow_index_dimensions(reduce_index<FlowIndex>(this->populations.size()))
+        , m_flow_range([&]() {
+            // take population dimensions, and set the compartment dimension to 1. this allows us to effectively
+            // iterate over all flow indices, but still have the corresponding PopIndex, where we can insert the
+            // flow source or target.
+            PopIndex dims   = this->populations.size();
+            get<Comp>(dims) = Index<Comp>{1};
+            return MultiIndexRange<PopIndex>(dims);
+        }())
+        , m_flow_values((this->populations.numel() / static_cast<size_t>(Comp::Count)) * Flows::size())
+    {
+    }
+
+    // LCOV_EXCL_START
+    // Exclude FLowModel::get_flows from coverage, as the function does nothing. It is intended to be overwritten by
+    // specific flow model implementations.
+
+    // Note: use get_flow_index when accessing flows
+    // Note: by convention, we compute incoming flows, thus entries in flows must be non-negative
+    virtual void get_flows(Eigen::Ref<const Eigen::VectorXd> /*pop*/, Eigen::Ref<const Eigen::VectorXd> /*y*/,
+                           double /*t*/, Eigen::Ref<Eigen::VectorXd> /*flows*/) const {};
+    // LCOV_EXCL_STOP
```

Comment:
To be honest, I copied this design pattern from CompartmentalModel::get_derivatives without too much thought. I think we can delete this default method, as we currently do not cast a Model to FlowModel anyways. But then the only hint as to how to implement get_flows would come from documentation, or other model implementations.

Reply:
It is a virtual function, so you could do `= 0`. Also, since we don't call `FlowModel` or `CompartmentModel` on its own, you should be able to just leave it declared (without virtual) but undefined. Since it isn't called, it shouldn't cause linker errors.

The `virtual` was added at the beginning, but we never actually used it.

Reply:
The comment in the CompartmentalModel states that we don't use a pure virtual function for easier python bindings. Do they still work with only a declaration? Should we then remove the virtual from both Models, and only use declarations?

Reply:
I don't fully understand it, but I think it has to do with binding the base class (`CompartmentModel<...>`) as `ModelBase` and then binding the actual Model class. I really don't think that's necessary, but I guess this is not the time or place to solve that.

Reply:
For now I only changed get_flows to a pure virtual function. Only using a declaration (in either model) did cause linker issues.